doc/develguide.rst

```diff
@@ -0,0 +1,656 @@
+          ASSERT_JVM_RUNNING("dosomething");
+
+          // Make a resource to capture any Java local references
+          JPJavaFrame frame;
+
+          // Call our Java methods
+          ...
+
+          // Return control to Python
+          return obj.keep();
+       }
+
+       // Use the standard catch to transfer any exceptions back
+       // to Python
+       PY_STANDARD_CATCH;
+
+       // Close out tracing
+       JP_TRACE_OUT;
+    }
+
+All entry points from Python into ``_jpype`` should be guarded with this pattern.
+
+There are exceptions to this pattern such as removing the logging, operating on 
+a call that does not need the jvm running, or operating where the frame is 
+already supported by the method being called.
+
+
+Python referencing
+~~~~~~~~~~~~~~~~~~
+
+One of the most miserable aspects of programming with CPython is the relative 
+inconsistancy of referencing.  Each method in Python may use a Python object or steal
+it, or it may return a borrowed reference or give a fresh reference.  Similar
+command such as getting an element from a list and getting an element from a tuple 
+can have different rules.  Because the was a constant source of bugs requiring 
+consultation of the Python manual for every line of code, we wrapped all of the
+Python calls we were required to work with in ``jp_pythontypes``.  
+
+Included in this wrapper is a Python reference counter called ``JPPyObject``.  
+Whenever an object is returned from Python it is immediately placed in smart 
+pointer ``JPPyObject`` with the policy that it was created with such as 
+``use_``, ``borrowed_``, ``claim_`` or ``call_``.
+
+``use_`` 
+  This policy means that the reference counter needs to be incremented and the start 
+  and the end.  We must reference it because it we don't and some Python call
+  destroys the refernce out from under us, the system may crash and burn.
+
+``borrowed_``
+  This policy means we were to be give a borrowed reference that we are expected
+  to reference and unreference when complete, but the command that returned it 
+  can fail.  Thus before reference it, the system must check if an error has
+  occurred.  If there is an error, it is promoted to an exception.
+
+``claim_``
+  This policy is used when we are given a new object with is already referenced
+  for us.  Thus we are to steal the reference for the duration of our use and
+  then dereference when we are done to keep it from leaking.
+
+``call_``
+  This policy both steals the reference and verifies there were no errors
+  prior to continuing.  Errors are promoted to exceptions when this reference 
+  is created.
+
+If we need to pass an object which is held in a smart pointer to Python
+which requires a reference, we call keep on the reference which transfers
+control to a ``PyObject*`` and prevents the pointer from removing the reference.
+As the object handle is leaving our control keep should only be called the
+return statement.
```

Comment:
This is very cool. Do we have rigorous memory leak checking? I can only think of one legacy test, which uses java.lang.Strings, but as you correctly mention, there are multiple sources of errors, when using the wrong mode in wrapping the Python objects.

Reply:
The leak checker code does a pretty good job of creating lots of java objects held by python and then destroyed.   If the code path hits the path of python wrapper layer in that case (tuples, dicts, etc) then it would have been caught.  We are a bit weaker for python leaks in general because I can't test the python wrapper directly.  I applied this smart pointer strategy because more often than not I messed something up.  Unfortunately, it only crashed if I undercounted references and overcounts just leak.